### DISCUSSION TITLE
Making Triangular and Polygonal Shapes in CSXCAD with AddPolygon and/or AddLinPoly with Python

### DISCUSSION THREAD

**Discussion by KS4VA** [thliebig/openEMS-Project#192] [Q&A Usage] [answered]
Hello all ... looking for some help on making 3D shapes that are not boxes and cylinders. Actually, I'd like to make some triangular extrusions. I'm trying to make sense of the AddPolygon and AddLinPoly functions. I can get them to run without error and appear as objects in AppCSXCAD, but they do not render at all. Exporting as STL and examining the "AddLinPoly" object shows just a line (not a 3D object), described as 8 vertices (mostly redundant). The exported STL for the AddPolygon is worse - an STL file with 8 duplicate vertices.

Has anyone been able to get this to work? The box is easy, here is some sample code and an AppCSXCAD screenshot:

`plate_metal = CSX.AddMetal('plate_9')
box9=plate_metal.AddBox([-25, -25, -25], [+25, +25, +25])

poly_metal = CSX.AddMetal('poly_1')
poly1=poly_metal.AddPolygon( ([-100,-100],[-10,-10]), "x", 50)

poly_metal = CSX.AddMetal('poly_2')
poly2=poly_metal.AddLinPoly( ([0,0],[200,200]), "z", 50, 100)`

 ____________

![AppCAD Poly](https://github.com/thliebig/openEMS-Project/assets/162363685/6f395101-5019-4b68-8ddc-8291471b1b9e)

**Comment by KS4VA** [comment, marked as answer]
Answering my own question here ... the syntax problem is in the way co-ordinate PAIRS are passed to the function. Answer came from

https://github.com/thliebig/openEMS-Project/discussions/56#discussion-5090881

It's not [ [x1,y1],[x2,y2],[x3,y3] ]  It should be [ [x1,x2,x3],[y1,y2,y3] ]

So a triangle could be either 

PolyMetal1 = CSX.AddMetal('PolyMetal1')
PolyMetal1.AddLinPoly(priority=10, points=[[0,100,0],[0,0,100]], norm_dir ='z', elevation=0, length=50)

or the same triangle in long form ...

    PolyMetal1 = CSX.AddMetal('PolyMetal1')
    pts_x = np.array([])
    pts_y = np.array([])
    pts_x = r_[pts_x, 0]
    pts_y = r_[pts_y, 0]
    pts_x = r_[pts_x, 100]
    pts_y = r_[pts_y, 0]
    pts_x = r_[pts_x, 0]
    pts_y = r_[pts_y, 100]
    pts = np.array([pts_x, pts_y])
    PolyMetal1.AddLinPoly(priority=10, points=pts, norm_dir ='z', elevation=0, length=50)

Hope this helps someone :)

**Comment by jackcarrozzo** [comment]
Just hit this, thank for the answer 😎
